### PR TITLE
Update README with instructions to run end-to-end tests on localmachine for release/1.2

### PR DESCRIPTION
**Pull request by scb01**

### CODE REVIEW

File: test/README.md
Comment (yophilav) [inline test/README.md:121]: I think this would be an opportunity to point to a corresponding doc: 
https://docs.microsoft.com/en-us/azure/iot-edge/how-to-auto-provision-symmetric-keys?view=iotedge-2020-11&tabs=windows

File: test/README.md
Comment (yophilav) [inline test/README.md:123]: Like wise for the DPS: 
https://docs.microsoft.com/en-us/azure/iot-edge/how-to-auto-provision-x509-certs?view=iotedge-2020-11&tabs=windows

File: test/README.md
Comment (yophilav) [inline test/README.md:184]: This typically a pain point to setup the new E2E. I wonder if 
1. We have a full template for context.json upload somewhere and that Cx can just download, fill in the blank and use it. 
2. Maybe we can have a script that generate context.json on the fly or a given set of tests?

File: test/README.md
Comment (yophilav) [inline test/README.md:274]: I wonder if we can script a really simple script that a user can execute so that the script will ask and source the password for them. That way there is no need to store a password in a plain text.

File: test/README.md
Comment (nimanch) [inline test/README.md:47]: Should we have a recommendation to use Microsoft Build of Open JDK since that is what we use to build code

https://docs.microsoft.com/en-us/java/openjdk/download

File: test/README.md
Comment (damonbarry) [inline test/README.md:268]: Earlier it says

> it is recommended that you create environment variables and make them available to the test framework in a way that avoids committing them to your shell's command history...

But then here, in our example script, we're creating environment variables in a way that commits them to our shell's command history (same for setting up the password with htpasswd earlier in the doc).

I usually use `read -s {var name}` to enter secrets on the command line, or `secret=$(az kv secret show ...)` if the values are in Key Vault. Both approaches keep secrets out of the command history. But I'm not sure the best way to do it here...

File: test/README.md
Comment (nimanch) [inline test/README.md:126]: Could we have more detailed instructions on this? Adding Sample commands would help

Comment (damonbarry) [inline test/README.md:126]: In the DPS section above, you wrote:

> A subset of the end to end tests will use DPS for testing device provisioning scenarios.

It would be good to have a similar explanation here for the storage account, since you don't need it at all if you skip the `TestUploadModuleLogs` and `TestUploadSupportBundle` tests.

File: test/README.md
Comment (scb01) [reply test/README.md:121]: Good point. Added.

File: test/README.md
Comment (scb01) [reply test/README.md:123]: Good point. Added.

File: test/README.md
Comment (scb01) [reply test/README.md:184]: For item 1,  I have a sample context.json further down in the readme that can be used as a starting point. 
For item 2,  I think this is a good idea. We should add that as a future work item.

File: test/README.md
Comment (scb01) [reply test/README.md:47]: Good point. Done.

File: test/README.md
Comment (scb01) [reply test/README.md:33]: We still have a dependency on aziot-identity-service which is in a separate repository. I'll add a TODO here to make sure that this is resolved in a later PR.

File: test/README.md
Comment (scb01) [reply test/README.md:110]: Yeah, that will make it easier to follow.

File: test/README.md
Comment (scb01) [reply test/README.md:115]: Addressed.

File: test/README.md
Comment (scb01) [reply test/README.md:134]: The mqttd build script is currently needed as part of the building of the container image for edge-hub. Skipping this step results in an error. I will add a TODO.

File: test/README.md
Comment (scb01) [reply test/README.md:268]: Good call out - our example should be `secure by default`. The best way is to probably get it setup in a secret store, but I am instead going to leave it up to the developer to do as they see fit. 

My go to mechanism, more out of convenience, is to create a script and run that in my setup. While this approach keeps the secrets out of the shell's command history, it does have the drawback of storing these secrets in the filesystem

File: test/README.md
Comment (damonbarry) [reply test/README.md:33]: Ah, right...I forgot you mentioned that earlier.

File: test/README.md
Comment (damonbarry) [reply test/README.md:268]: I agree our example should be secure by default. How best to balance that with simplicity, ease of use, etc. is your call...

File: test/README.md
Comment (damonbarry) [reply test/README.md:126]: Perhaps we could just point to some docs about this, similar to creating hub and DPS? But I'm not sure if the right doc exists (create a storage container, set access level to private, grab the sas token URL).

Comment (scb01) [reply test/README.md:126]: Good idea.

Comment (scb01) [reply test/README.md:126]: I will spend some time with the CLI to see if all of this can be setup from the command line. Once I have it working, I'll replace (or add) the commands. This may be in a later PR.

File: test/README.md
Comment (scb01) [reply test/README.md:186]: I've put this into a separate file. Lets see if this flows better.

File: test/README.md
Comment (nlcamp) [inline test/README.md]: Do we just create a root CA cert and that's it for now? Is there another implied step? If it will be used later in the procedure, it would be good to mention that here.

File: test/README.md
Comment (nlcamp) [inline test/README.md:72]: Should we instead encourage install of the Moby engine? https://docs.microsoft.com/en-us/azure/iot-edge/how-to-install-iot-edge?view=iotedge-2020-11#install-a-container-engine

Comment (scb01) [reply test/README.md]: Good callout. Done.

File: test/README.md
Comment (damonbarry) [reply test/README.md]: I see the TODO was added, resolving...

File: test/README.md
Comment (damonbarry) [reply test/README.md]: @scb01 I still see several references to "IoTHub" (no space)

File: test/README.md
Comment (damonbarry) [inline test/README.md:254]: Can you indent the body of this JSON object? Also, can you do 4-space indent instead of 8-space?

File: test/doc/end-to-end-test-config.md
Comment (damonbarry) [inline test/doc/end-to-end-test-config.md:1]: This title isn't rendering correctly...I think you need a space after the '#'

File: test/doc/end-to-end-test-config.md
Comment (damonbarry) [inline test/doc/end-to-end-test-config.md:38]: `E2E_DPS_GROUP_KEY` should have an '*' not a '✔' in the Required column, since it's only required for DPS tests, not all tests.

Comment (damonbarry) [reply test/doc/end-to-end-test-config.md:38]: ...although I see several mistakes in the Required column, and probably none of them are from you. :) Perhaps we need to make a run the the Required column as a separate effort.

File: test/README.md
Comment (damonbarry) [inline test/README.md:262]: This link URI is missing the '.md' extension...does it still work?

File: test/README.md
Comment (damonbarry) [inline test/README.md:221]: This link URI (`./doc/end-to-end-test-config`) is missing the '.md' extension...does it still work?

File: test/README.md
Comment (damonbarry) [reply test/README.md]: Looks good

File: test/README.md
Comment (nlcamp) [inline test/README.md:81]: Are there any suggestions for which fields to provide when prompted after running this command?

File: test/README.md
Comment (nlcamp) [inline test/README.md:73]: Can any of the test certs generated in the earlier step be used here? If not, can our certGen.sh script be used to generate the certs in this step? If so, what args should be passed? If not, is this functionality something we want to add to certGen.sh in the future (i.e. should it be on our backlog)?

File: test/README.md
Comment (nlcamp) [inline test/README.md:124]: Are we actually supposed to run  `sudo docker login localhost:5000` here and then specify the username and password that was set above in the "Setup native basic auth" step? If so, it would be good to make this explicit here. If not, it would be good to explain which registry we're logging into here.

File: test/README.md
Comment (nlcamp) [inline test/README.md:100]: nit: for even more clarity and consistency, consider changing _container repository_ to _container registry_ here and throughout the doc.

File: test/README.md
Comment (nlcamp) [inline test/README.md:133]: I had to run `chmod u+x dotnet-install.sh` before this command.

File: test/README.md
Comment (nlcamp) [inline test/README.md:139]: 1. Install the Azure CLI. The second bullet under 'Prerequisites' in the IoT Edge Quickstart document has instructions: https://docs.microsoft.com/en-us/azure/iot-edge/quickstart-linux?view=iotedge-2020-11#prerequisites
2. Create a resource group: `az group create --name IoTEdgeResources --location westus2`
3. Create an IoT hub: `az iot hub create --resource-group IoTEdgeResources --name {hub_name} --sku F1 --partition-count 2`
4. Get the IoT hub connection string: `hubConnectionString=$(az iot hub show-connection-string --name {hub_name} --key primary --query connectionString -o tsv)`
5. Create a Device Provisioning Service:  `az iot dps create --name {dps_name} --resource-group IoTEdgeResources --location westus`
6. Link the IoT hub and dps: `az iot dps linked-hub create --dps-name {dps_name} --resource-group IoTEdgeResources --connection-string $hubConnectionString --location westus`
7. Create an Azure Storage Account: `az storage account create  --name <account-name>  --resource-group IoTEdgeResources  --location westus  --sku Standard_RAGRS  --kind StorageV2`

File: test/README.md
Comment (scb01) [reply test/README.md:124]: Good catch. That is exactly what is needed here. I'll address this in the PR.

File: test/README.md
Comment (scb01) [reply test/README.md:73]: IMO, these are unrelated. The certGen scripts are to create certs for device identity and communication, while these are TLS for the container registry. We should keep them separated out.

File: test/README.md
Comment (scb01) [reply test/README.md:133]: I recall I had to do something similar. Will add.

File: test/README.md
Comment (scb01) [reply test/README.md:100]: Will do.

File: test/doc/end-to-end-test-config.md
Comment (scb01) [reply test/doc/end-to-end-test-config.md:38]: Wasn't sure what the semantic difference was between the ✔ and the *.  I'll document that in the PR i.e, ✔ means required and * means conditionally required. 

Also, I think we should take this opportunity to clean up the Required column. I did find that to be pretty confusing as I was setting up my environment.

File: test/doc/end-to-end-test-config.md
Comment (scb01) [reply test/doc/end-to-end-test-config.md:1]: Will do.

File: test/README.md
Comment (scb01) [reply test/README.md:254]: Will do

File: test/README.md
Comment (scb01) [reply test/README.md:262]: It was working with or without the .md extension in vscode. I'll try this out on my fork and validate that it working.

File: test/doc/end-to-end-test-config.md
Comment (damonbarry) [reply test/doc/end-to-end-test-config.md:38]: Yep, agreed...anything we can do to make this easier to read and understand is a win.

Comment (scb01) [reply test/doc/end-to-end-test-config.md:38]: Changed the ✔and added a blurb about the difference between * and ✔

File: test/README.md
Comment (scb01) [reply test/README.md:81]: I found the prompts to be self explanatory ( and also standard as this the type of info you would provide to create a CSR in pretty much all contexts). I don't feel strongly about it( except to may be keep the content of this long doc as short as possible).  Happy to add if you if you think it would be helpful.

File: test/README.md
Comment (scb01) [reply test/README.md]: Validated that it needs the .md. Added.

File: test/README.md
Comment (scb01) [reply test/README.md]: Validated that it needs the `.md`. Added.

File: test/README.md
Comment (nlcamp) [inline test/README.md:171]: nit: forgot the .sh (i.e. buildBranch.sh)

Comment (scb01) [reply test/README.md:171]: Nice catch! Addressed.

File: test/README.md
Comment (scb01) [reply test/README.md]: I have taken the details on how to set the environment out of the README and left it to the developer's discretion.

File: test/README.md
Comment (nlcamp) [reply test/README.md:141]: 7a. Retrieve the connection string for the storage account: `sa_conn_string=$(az storage account show-connection-string --name {storage-account-name} --resource-group IoTEdgeResources -o tsv)`
7b. Create a storage container: `az storage container create --name {container-name} --resource-group IoTEdgeResources --account-name {storage-account-name} --connection-string $sa_conn_string`
7c. Generate a SAS token for the storage container: `sas=$(az storage container generate-sas --account-name {storage-account-name} --name {container-name} --https-only --permissions w --connection-string $sa_conn_string)`

File: test/README.md
Comment (nlcamp) [inline test/README.md:223]: Should the path in this section say `netcoreapp3.1` instead of `netcoreapp2.1`?

Comment (nlcamp) [inline test/README.md:223]: The 'end-to-end test parameters' link returned a 404 error for me. I think it should go to `./doc/e2e.md#test-parameters` instead?

File: test/README.md
Comment (scb01) [reply test/README.md:141]: Thanks. I'll add these to the PR.

File: test/README.md
Comment (scb01) [reply test/README.md:223]: Oh, sharp eyes! Yep, 3.1 is right. I will fix it.

Comment (scb01) [reply test/README.md:223]: I don't think they render properly on the PR. I tried it on my fork and it rendered fine.  This should be good once the PR is merged.

File: test/README.md
Comment (damonbarry) [inline test/README.md:152]: @scb01 There can only be one free IoT hub per subscription, and the quotas are extremely tight so it won't take long to run into throttling. I wonder if recommending the free SKU here might generate support issues that are unrelated to the tests...

File: test/README.md
Comment (damonbarry) [reply test/README.md]: @scb01 There are still four instances of "iothub" (no space, with various casing). Can you correct these?

File: test/README.md
Comment (scb01) [reply test/README.md:152]: I agree. While the intent was to provide a cost effective option, it may result in confusion due to the tight quotas. I will go without specifying the SKU or the partition count ( which defaults to the SKU S1 and 4 respectively).  The idea of removing the partition count is that decreases the concept count by one and reduces the cognitive burden.

File: test/README.md
Comment (nlcamp) [inline test/README.md:73]: Although it's arguably obvious, it would help to have a statement like "Be sure to update the parameter values from the sample context.json to values that are valid for your environment". I ran into an issue where I was using the wrong dpsScopeId and it took a while to troubleshoot.

Comment (scb01) [reply test/README.md:73]: Doesn't hurt to specify this. Done.

File: test/doc/one-time-setup.md
Comment (nlcamp) [inline test/doc/one-time-setup.md:189]: `--verified` is not a valid arg in the version of Azure CLI and/or IoT extension that I'm using.

File: test/README.md
Comment (nlcamp) [reply test/README.md]: Fair enough.

File: test/README.md
Comment (nlcamp) [reply test/README.md]: Oh, I see. I missed the fact that below you're changing the name from e2e.md to end-to-end-test-config.md

File: test/doc/one-time-setup.md
Comment (nlcamp) [reply test/doc/one-time-setup.md:189]: Resolved by updating my version of Azure CLI and the iot extension (both were out of date).

File: test/README.md
Comment (scb01) [reply test/README.md]: This should be good to go now.

File: test/README.md
Comment (scb01) [reply test/README.md]: Instructions updated with CLI commands.

**Review by nlcamp — APPROVED** [review]
This is a great doc!

**Review by nlcamp — APPROVED** [review]

**Pull request merged**
